### PR TITLE
core/mount: mount command may fail after adding the corresponding proc mountinfo entry

### LINKED ISSUE #17617: Mount unit enters active (mounted) state when mount process exits with status=1/FAILURE
**systemd version the issue has been seen with**
 > Systemd 246

**Used distribution**
 > Debian Bullseye

**Linux kernel version used**
 > `5.8.0-2-amd64 #1 SMP Debian 5.8.10-1`

**CPU architecture issue was seen on**
 > amd64

**Expected behaviour you didn't see**
 > When the mount process exits with a non-zero status, I would expect the mount unit to enter failed state accordingly. Interestingly, when an unmount fails (eg `target is busy` which returns `status=32/n/a`) it does register as failed: the next line in the log reads `"<unitname>.mount: Failed with result 'exit-code'.`, and the `OnFailure` dependency is triggered. I honestly don't know whether it's intended/expected that a failed attempt to unmount a filesystem that's not mounted causes the mount unit to enter a failed state, but I would certainly expect a failed *mount* process to do so.

**Unexpected behaviour you saw**
 > When the mount process returns  `status=1/FAILURE`, the mount unit enters state `active (mounted)` and the `OnFailure` dependency is not triggered (but the filesystem is of course not mounted). I found this particularly surprising given that the log shown by journalctl indicates the mount process exit status as "FAILURE".

**Steps to reproduce the problem**
 > In my case, it's an sshfs mount triggered by an automount unit. Making the ssh key inaccessible will cause the mount process to fail (`Connection reset by peer`), triggering the bug. I have not tested with other filesystems, or with other failure modes yielding different exit codes than the ones mentioned above.

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #17617.

(I have not tested this yet.)

### CODE REVIEW

**Comment by raven-au** [comment]
> Hopefully fixes #17617.
> 
> (I have not tested this yet.)

And likely #17545.

**Comment by raven-au** [comment]
I'm sorry, I don't quite get how this commit works.
Can you explain your thinking on the state processing here please.

There are a couple of things that confuse me.

The statement "adding the corresponding proc mountinfo" is misleading because proc mountinto is generated from the kernel mount table structures on access, there's no adding of entries as such.

Is this a little racy in that the SIGCHLD and events received when the proc mounts changes both call mount_process_proc_self_mountinfo() so how does the state end up being correctly processed, is there a possible call ordering race?

**Comment by yuwata** [comment]
@raven-au When a .mount unit is successfully started, the state of the unit becomes `MOUNT_MOUNTING` -> `MOUNT_MOUNTING_DONE` -> `MOUNT_MOUNTED`. More specifically, `mount_start()` invokes `mount` command and sets the state to `MOUNT_MOUNTING` first. Then, before `mount` command exits, by a libmount event, `mount_process_proc_self_mountinfo()` sets the state to `MOUNT_MOUNTING_DONE`. Finally, when `mount` command finished, in `mount_sigchld()` sets the state to `MOUNT_MOUNTED`.

- `MOUNT_MOUNTING`: `mount` command is running, and the corresponding mountinfo entry has not been added yet.
- `MOUNT_MOUNTING_DONE`: `mount` command is still running, and the mountinfo entry has already added.
- `MOUNT_MOUNTED`: `mount` command successfully finished.

However, I have not checked the code of fuse or something corresponding though, if once `mount` command makes kernel to add the mountinfo entry (thus state is changed `MOUNT_MOUNTING` -> `MOUNT_MOUNTING_DONE`), and then fails with some reason, and removes the mountinfo entry, previously `mount_process_proc_self_mountinfo()` does not change the state and it is kept as `MOUNT_MOUNTING_DONE`. Thus, even if the subsequent `mount_sigchld()` detects the failure, the state is changed `MOUNT_MOUNTING_DONE` -> `MOUNT_MOUNTED`.

This patch makes `mount_process_proc_self_mountinfo()` set the state back to `MOUNT_MOUNTING_DONE` -> `MOUNT_MOUNTING` when the corresponding mountinfo entry is removed due to an error in `mount` command. Thus, the subsequent `mount_sigchld()` correctly changes the state to `MOUNT_MOUNTING` -> `MOUNT_FAILED` with `mount_enter_dead()`.

**Comment by yuwata** [comment]
> > Hopefully fixes #17617.
> > (I have not tested this yet.)
> 
> And likely #17545.

I am not sure about #17545. I cannot follow what is the problem in that issue. It sounds user's misconfiguration for me. No?

**Comment by raven-au** [comment]
> > > Hopefully fixes #17617.
> > > (I have not tested this yet.)
> > 
> > 
> > And likely #17545.
> 
> I am not sure about #17545. I cannot follow what is the problem in that issue. It sounds user's misconfiguration for me. No?

Basically yes, but I'm pretty sure it's the same as the problem your trying to fix here.

I hadn't looked into it yet but I was thinking it was an incorrect return from mount causing systemd to think the mount had succeeded when it hadn't but having seen this I think it is the same issue.

**Comment by raven-au** [comment]
> @raven-au When a .mount unit is successfully started, the state of the unit becomes `MOUNT_MOUNTING` -> `MOUNT_MOUNTING_DONE` -> `MOUNT_MOUNTED`. More specifically, `mount_start()` invokes `mount` command and sets the state to `MOUNT_MOUNTING` first. Then, before `mount` command exits, by a libmount event, `mount_process_proc_self_mountinfo()` sets the state to `MOUNT_MOUNTING_DONE`. Finally, when `mount` command finished, in `mount_sigchld()` sets the state to `MOUNT_MOUNTED`.
> 
>     * `MOUNT_MOUNTING`: `mount` command is running, and the corresponding mountinfo entry has not been added yet.
> 
>     * `MOUNT_MOUNTING_DONE`: `mount` command is still running, and the mountinfo entry has already added.
> 
>     * `MOUNT_MOUNTED`: `mount` command successfully finished.

Thanks for taking the time to describe this.

> 
> 
> However, I have not checked the code of fuse or something corresponding though, if once `mount` command makes kernel to add the mountinfo entry (thus state is changed `MOUNT_MOUNTING` -> `MOUNT_MOUNTING_DONE`), and then fails with some reason, and removes the mountinfo entry, previously `mount_process_proc_self_mountinfo()` does not change the state and it is kept as `MOUNT_MOUNTING_DONE`. Thus, even if the subsequent `mount_sigchld()` detects the failure, the state is changed `MOUNT_MOUNTING_DONE` -> `MOUNT_MOUNTED`.

I had a quick look at the sshfs code and it perform the fuse mount first then raises the ssh connection to get a socket connection.
If the ssh connection fails it umounts the fuse mount so your analysis of what's happening looks accurate.

> 
> This patch makes `mount_process_proc_self_mountinfo()` set the state back to `MOUNT_MOUNTING_DONE` -> `MOUNT_MOUNTING` when the corresponding mountinfo entry is removed due to an error in `mount` command. Thus, the subsequent `mount_sigchld()` correctly changes the state to `MOUNT_MOUNTING` -> `MOUNT_FAILED` with `mount_enter_dead()`.

I see, so your relying on the inherent single threaded'ness of the event system to not be interrupted between setting this in mount_process_proc_self_mountinfo() of the sigchld handler and its subsequent state processing, right?

**Comment by raven-au** [comment]
> > > 
> > > And likely #17545.
> > 
> > 
> > I am not sure about #17545. I cannot follow what is the problem in that issue. It sounds user's misconfiguration for me. No?
> 
> Basically yes, but I'm pretty sure it's the same as the problem your trying to fix here.
> 
> I hadn't looked into it yet but I was thinking it was an incorrect return from mount causing systemd to think the mount had succeeded when it hadn't but having seen this I think it is the same issue.

The problem in #17545 ended up being the target host key not being present in the root user known_hosts file which caused the ssh connection to fail, which is the same symptom as is seen here.

**Comment by yuwata** [comment]
> > However, I have not checked the code of fuse or something corresponding though, if once `mount` command makes kernel to add the mountinfo entry (thus state is changed `MOUNT_MOUNTING` -> `MOUNT_MOUNTING_DONE`), and then fails with some reason, and removes the mountinfo entry, previously `mount_process_proc_self_mountinfo()` does not change the state and it is kept as `MOUNT_MOUNTING_DONE`. Thus, even if the subsequent `mount_sigchld()` detects the failure, the state is changed `MOUNT_MOUNTING_DONE` -> `MOUNT_MOUNTED`.
> 
> I had a quick look at the sshfs code and it performs the fuse mount first then raises the ssh connection to get a socket connection.
> If the ssh connection fails it umounts the fuse mount so your analysis of what's happening looks accurate.

Thanks!

> > This patch makes `mount_process_proc_self_mountinfo()` set the state back to `MOUNT_MOUNTING_DONE` -> `MOUNT_MOUNTING` when the corresponding mountinfo entry is removed due to an error in `mount` command. Thus, the subsequent `mount_sigchld()` correctly changes the state to `MOUNT_MOUNTING` -> `MOUNT_FAILED` with `mount_enter_dead()`.
> 
> I see, so your relying on the inherent single threaded'ness of the event system to not be interrupted between setting this in mount_process_proc_self_mountinfo() of the sigchld handler and its subsequent state processing, right?

Right.

**Comment by raven-au** [comment]
Then the only other thing that I wonder about is whether this would affect proc mounts processing for mounts done outside of systemd which it tracks via proc mounts change notifications.
I guess I'll need to look at the code to work that out, ;)

**Comment by yuwata** [comment]
When the `mount` command is directly invoked by a user (and when it succeeds), the state of the corresponding .mount unit directly changes `MOUNT_DEAD` -> `MOUNT_MOUNTED` in `mount_process_proc_mountinfo()`.

When fuse.sshfs fails, then it should be `MOUNT_DEAD` -> `MOUNT_MOUNTED` -> `MOUNT_DEAD`. As the corresponding mountinfo is added first, and then removed. Note that even if `mount` command fails, the .mount unit does not enter failed state. But this is consistent to the case of failure in 'normal' mount, as it does not update mountinfo, so the libmount event is not triggered, thus, the state is still in `MOUNT_DEAD`.

**Comment by raven-au** [comment]
Thanks very much for the effort describing the mount state behaviours.
That will be helpful to me going forward, ;)

**Comment by raven-au** [comment]
> Thanks very much for the effort describing the mount state behaviours.
> That will be helpful to me going forward, ;)

I could review the change if you need me too but I've really only tested this rather than looked closely at the source behaviours.
Let me know if you need me to do anything in support of this change.

**Comment by yuwata** [comment]
> I could review the change if you need me too but I've really only tested this rather than looked closely at the source behaviours.
> Let me know if you need me to do anything in support of this change.

@raven-au Any kind of helps are always welcome.

**Comment by raven-au** [comment]
> > I could review the change if you need me too but I've really only tested this rather than looked closely at the source behaviours.
> > Let me know if you need me to do anything in support of this change.
> 
> @raven-au Any kind of helps are always welcome.

I'm sorry, but after reviewing the code I find the approach used a bit counter intuitive.

The handling of the success or failure of a mount command is really the job of the SIGCHLD handler and setting the mount state in the proc mountinfo handling to influence the SIGCHLD processing seems like it's the wrong way to fix this.

I think the right place to check for and handle the exec failure is actually in the SIGCHLD handler.

This patch seems to work too.

`commit 626c81e19dd2894d945f1246178a6e24df430882
Author: Ian Kent <raven@themaw.net>
Date:   Sun Nov 22 07:42:35 2020 +0800

    mount: fix failed mount detection in sigchld processing
    
    The SIGCHLD processing of a mount failes to take account of the case
    where a mount can be set to MOUNT_MOUNTING_DONE via the proc mountinfo
    processing but the mount command ultimately fails.
    
    Fix this by checking for a mount command failure in the SIGCHLD case
    processing of MOUNT_MOUNTING_DONE.

diff --git a/src/core/mount.c b/src/core/mount.c
index faa2cb0e03..3053a75240 100644
--- a/src/core/mount.c
+++ b/src/core/mount.c
@@ -1402,7 +1402,11 @@ static void mount_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                 break;
 
         case MOUNT_MOUNTING_DONE:
-                mount_enter_mounted(m, f);
+                /* The mount may have appeared in the mount table but ultimately failed. */
+                if (f == MOUNT_SUCCESS)
+                        mount_enter_mounted(m, f);
+                else
+                        mount_enter_dead(m, f);
                 break;
 
         case MOUNT_REMOUNTING:`

**Comment by yuwata** [comment]
> The handling of the success or failure of a mount command is really the job of the SIGCHLD handler and setting the mount state in the proc mountinfo handling to influence the SIGCHLD processing seems like it's the wrong way to fix this.
> 
> I think the right place to check for and handle the exec failure is actually in the SIGCHLD handler.
> 
> This patch seems to work too.

@raven-au Yes, your patch should also work. And, actually I initially thought to use your approach. However, at least in the current logic, pid1 honors the entries in /proc/self/mountinfo rather than the exit status of `mount` command. That is, even if `mount` command fails, the corresponding mount entry exists in /proc/self/mountinfo, then pid1 treats that the .mount unit should be in mounted state. Or, even if `mount` command succeeds, the corresponding mount does not exist in /proc/self/mountinfo, then the .mount unit enters failed state. See the comment in `mount_sigchld()`:
```
        /* Note that due to the io event priority logic, we can be sure the new mountinfo is loaded
         * before we process the SIGCHLD for the mount command. */
```
Also,
```
        case MOUNT_MOUNTING:
                /* Our mount point has not appeared in mountinfo.  Something went wrong. */

                if (f == MOUNT_SUCCESS) {
                        /* Either /bin/mount has an unexpected definition of success,
                         * or someone raced us and we lost. */
                        log_unit_warning(UNIT(m), "Mount process finished, but there is no mount.");
                        f = MOUNT_FAILURE_PROTOCOL;
                }
                mount_enter_dead(m, f);
                break;
```
So, in the case of the issue #17617, mount state should be correctly updated based on /proc/self/mountinfo, I think.

**Comment by raven-au** [comment]
> @raven-au Yes, your patch should also work. And, actually I initially thought to use your approach. However, at least in the current logic, pid1 honors the entries in /proc/self/mountinfo rather than the exit status of `mount` command. That is, even if `mount` command fails, the corresponding mount entry exists in /proc/self/mountinfo, then pid1 treats that the .mount unit should be in mounted state. Or, even if `mount` command succeeds, the corresponding mount does not exist in /proc/self/mountinfo, then the .mount unit enters failed state. See the comment in `mount_sigchld()`:

Arrrh yes.

Your saying you don't want to have to trust the execed command.
Sadly that type of problem has happened to me all too often with autofs so I agree with your reasoning.

snip ...
 
> So, in the case of the issue #17617, mount state should be correctly updated based on /proc/self/mountinfo, I think.

Yes, that does make sense even though it looks unintuitive, it caught me out, ;)

File: src/core/mount.c [hunk 1]
```diff
@@ -1948,6 +1948,16 @@ static int mount_process_proc_self_mountinfo(Manager *m) {
...
                                 break;
 
+                        case MOUNT_MOUNTING_DONE:
+                                /* The mount command may once add the corresponding proc mountinfo
+                                 * entry and then remove it because of an internal error. E.g.,
+                                 * fuse.sshfs seems to do that when the connection fails. See #17617.
+                                 * To handle such the case, let's once set the state back to mounting.
+                                 * Then, the unit can correctly enter the failed state later in
+                                 * mount_sigchld(). */
+                                mount_set_state(mount, MOUNT_MOUNTING);
```
Comment (raven-au) [inline src/core/mount.c:1957]: Small nit.
s/mount command may once add/mount command may add/
is clearer I think.

**Review by raven-au — APPROVED** [review]
Looks good to me and as discussed this does appear to be the right way to resolve the problem.
That wording nit is only very minor so is not really important.

**Comment by yuwata** [comment]
@raven-au Thank you for the review. The comment in the code is updated.

> Your saying you don't want to have to trust the execed command.

Right.

**Comment by poettering** [comment]
lgtm

**Pull request merged**